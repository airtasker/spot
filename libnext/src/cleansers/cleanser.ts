import { ContractDefinition } from "../models/definitions";
import { ContractNode } from "../models/nodes";
import { cleanseApi } from "./nodes/api-cleanser";
import { cleanseEndpoint } from "./nodes/endpoint-cleanser";

/**
 * Removes all additional parsing information associated with a parsed contract.
 *
 * @param contractNode a parsed contract
 */
export function cleanse(contractNode: ContractNode): ContractDefinition {
  const api = cleanseApi(contractNode.api.value);
  const endpoints = contractNode.endpoints.map(endpoint =>
    cleanseEndpoint(endpoint.value)
  );
  const types = contractNode.types;

  return { api, endpoints, types };
}
