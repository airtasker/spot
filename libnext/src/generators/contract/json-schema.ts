import assertNever from "assert-never";
import * as YAML from "js-yaml";
import { ContractDefinition } from "libnext/src/models/definitions";
import { DataType, TypeKind } from "../../models/types";
import compact = require("lodash/compact");

export function generateJsonSchema(
  contractDefinition: ContractDefinition,
  format: "json" | "yaml"
) {
  const contract = jsonSchema(contractDefinition);
  switch (format) {
    case "json":
      return JSON.stringify(contract, null, 2);
    case "yaml":
      return YAML.safeDump(contract);
    default:
      throw assertNever(format);
  }
}

export function jsonSchema(contractDefinition: ContractDefinition): JsonSchema {
  return {
    $schema: "http://json-schema.org/draft-07/schema#",
    definitions: contractDefinition.types.reduce<{
      [typeName: string]: JsonSchemaType;
    }>((acc, typeNode) => {
      acc[typeNode.name] = jsonTypeSchema(typeNode.type);
      return acc;
    }, {})
  };
}

export function jsonTypeSchema(type: DataType): JsonSchemaType {
  switch (type.kind) {
    case TypeKind.NULL:
      return {
        type: "null"
      };
    case TypeKind.BOOLEAN:
      return {
        type: "boolean"
      };
    case TypeKind.BOOLEAN_LITERAL:
      return {
        type: "boolean",
        const: type.value
      };
    case TypeKind.DATE:
    case TypeKind.DATE_TIME:
    case TypeKind.STRING:
      return {
        type: "string"
      };
    case TypeKind.STRING_LITERAL:
      return {
        type: "string",
        const: type.value
      };
    case TypeKind.NUMBER:
      return {
        type: "number"
      };
    case TypeKind.INTEGER:
      return {
        type: "integer"
      };
    case TypeKind.NUMBER_LITERAL:
      return Math.round(type.value) === type.value
        ? {
            type: "integer",
            const: type.value
          }
        : {
            type: "number",
            const: type.value
          };
    case TypeKind.OBJECT:
      return type.properties.reduce<JsonSchemaObject & { required: string[] }>(
        (acc, property) => {
          if (!property.optional) {
            acc.required.push(property.name);
          }
          acc.properties[property.name] = jsonTypeSchema(property.type);
          return acc;
        },
        {
          type: "object",
          properties: {},
          required: []
        }
      );
    case TypeKind.ARRAY:
      return {
        type: "array",
        items: jsonTypeSchema(type.elements)
      };
    case TypeKind.UNION:
      return {
        oneOf: type.types.map(jsonTypeSchema)
      };
    case TypeKind.TYPE_REFERENCE:
      return {
        $ref: `#/definitions/${type.name}`
      };
    default:
      throw assertNever(type);
  }
}

export interface JsonSchema {
  $schema: "http://json-schema.org/draft-07/schema#";
  definitions: {
    [typeName: string]: JsonSchemaType;
  };
}

export type JsonSchemaType =
  | JsonSchemaObject
  | JsonSchemaArray
  | JsonSchemaOneOf
  | JsonSchemaNull
  | JsonSchemaString
  | JsonSchemaNumber
  | JsonSchemaInteger
  | JsonSchemaBoolean
  | JsonSchemaTypeReference;

export interface JsonSchemaObject {
  type: "object";
  properties: {
    [name: string]: JsonSchemaType;
  };
  required?: string[];
}

export interface JsonSchemaArray {
  type: "array";
  items: JsonSchemaType;
}

export interface JsonSchemaOneOf {
  oneOf: JsonSchemaType[];
  discriminator?: {
    propertyName: string;
    mapping: {
      [value: string]: JsonSchemaType;
    };
  };
}

export interface JsonSchemaNull {
  type: "null";
}

export interface JsonSchemaString {
  type: "string";
  const?: string;
  enum?: string[];
}

export interface JsonSchemaNumber {
  type: "number";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaInteger {
  type: "integer";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaBoolean {
  type: "boolean";
  const?: boolean;
}

export interface JsonSchemaTypeReference {
  $ref: string;
}
