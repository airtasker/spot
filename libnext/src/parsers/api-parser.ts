import { ClassDeclaration } from "ts-simple-ast";
import {
  extractJsDocComment,
  extractStringProperty,
  extractDecoratorFactoryConfiguration
} from "./utilities/parser-utility";
import { ApiDefinition } from "../models/definitions";

/**
 * Parse an `@api` decorated class.
 *
 * @param klass a class declaration
 */
export function parseApi(klass: ClassDeclaration): ApiDefinition {
  const decorator = klass.getDecoratorOrThrow("api");
  const description = extractJsDocComment(klass);
  const configuration = extractDecoratorFactoryConfiguration(decorator);
  const name = extractStringProperty(configuration, "name");
  return { name, description };
}
