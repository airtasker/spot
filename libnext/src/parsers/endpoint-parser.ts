import { ClassDeclaration, ObjectLiteralExpression } from "ts-simple-ast";
import {
  extractJsDocComment,
  extractDecoratorFactoryConfiguration,
  extractStringProperty,
  isHttpMethod
} from "./parser-utility";
import { EndpointDefinition } from "../models/definitions";
import { HttpMethod } from "../models/types";

/**
 * Parse an Endpoint definition object.
 *
 * @param klass a class declaration
 */
export function parseEndpoint(klass: ClassDeclaration): EndpointDefinition {
  const decorator = klass.getDecoratorOrThrow("endpoint");
  const description = extractJsDocComment(klass);
  const configuration = extractDecoratorFactoryConfiguration(decorator);
  const method = extractHttpMethodProperty(configuration, "method");
  const path = extractStringProperty(configuration, "path");
  return { description, method, path };
}

/**
 * Extract a HTTP method property from an object literal.
 *
 * @param objectLiteral an object literal
 * @param propertyName the property to extract
 */
function extractHttpMethodProperty(
  objectLiteral: ObjectLiteralExpression,
  propertyName: string
): HttpMethod {
  const method = extractStringProperty(objectLiteral, propertyName);
  if (!isHttpMethod(method)) {
    throw new Error(`expected a HttpMethod, got ${method}`);
  }
  return method;
}
