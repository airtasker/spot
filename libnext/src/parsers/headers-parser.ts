import { HeaderDefinition } from "../models/definitions";
import { ParameterDeclaration } from "ts-simple-ast";
import {
  ensureNodeNotOptional,
  extractObjectParameterProperties,
  ensureDataTypeIsKind,
  extractJsDocComment
} from "./utilities/parser-utility";
import { parseType } from "./utilities/type-parser";
import { KindOfString, KindOfNumber } from "../models/types";

/**
 * Parse an `@headers` decorated parameter.
 *
 * @param parameter a parameter declaration
 */
export function parseHeaders(
  parameter: ParameterDeclaration
): HeaderDefinition[] {
  parameter.getDecoratorOrThrow("headers");
  ensureNodeNotOptional(parameter);
  const properties = extractObjectParameterProperties(parameter);
  return properties.map(property => {
    const propertyDataType = parseType(property.getType());
    ensureDataTypeIsKind(propertyDataType, KindOfString.concat(KindOfNumber));
    return {
      name: property.getName(),
      description: extractJsDocComment(property),
      type: propertyDataType,
      optional: property.hasQuestionToken()
    };
  });
}
