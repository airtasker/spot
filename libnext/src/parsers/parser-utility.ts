import {
  ts,
  ObjectLiteralExpression,
  JSDocableNode,
  Decorator,
  TypeGuards
} from "ts-simple-ast";
import { HttpMethod } from "../../../lib/src/lib";

/**
 * Extracts the JS Doc comment from a node.
 *
 * @param node a JS Docable Node
 */
export function extractJsDocComment(node: JSDocableNode): string | undefined {
  const jsDocs = node.getJsDocs();
  if (jsDocs.length === 1) {
    return jsDocs[0].getComment();
  } else if (jsDocs.length > 1) {
    throw new Error(`expected 1 jsDoc node, got ${jsDocs.length}`);
  }
  return;
}

/**
 * Extract a string property from an object literal.
 *
 * @param objectLiteral an object literal
 * @param propertyName the property to extract
 */
export function extractStringProperty(
  objectLiteral: ObjectLiteralExpression,
  propertyName: string
): string {
  return objectLiteral
    .getPropertyOrThrow(propertyName)
    .getLastChildIfKindOrThrow(ts.SyntaxKind.StringLiteral)
    .getLiteralText();
}

/**
 * Extract an optional string property from an object literal.
 *
 * @param objectLiteral an object literal
 * @param propertyName the property to extract
 */
export function extractOptionalStringProperty(
  objectLiteral: ObjectLiteralExpression,
  propertyName: string
): string | undefined {
  const property = objectLiteral.getProperty(propertyName);
  if (property) {
    return property
      .getLastChildIfKindOrThrow(ts.SyntaxKind.StringLiteral)
      .getLiteralText();
  }
  return;
}

/**
 * Extract the Configuration object (first argument) from a Spot decorator factory.
 *
 * @param decorator a decorator factory
 */
export function extractDecoratorFactoryConfiguration(
  decorator: Decorator
): ObjectLiteralExpression {
  if (decorator.getArguments().length === 1) {
    const argument = decorator.getArguments()[0];
    if (TypeGuards.isObjectLiteralExpression(argument)) {
      return argument;
    } else {
      throw new Error(`expected object literal`);
    }
  } else {
    throw new Error(
      `expected 1 argument, got ${decorator.getArguments().length}`
    );
  }
}

/**
 * Determine if a string is a HTTP method.
 *
 * @param method the string
 */
export function isHttpMethod(method: string): method is HttpMethod {
  switch (method) {
    case "GET":
    case "HEAD":
    case "POST":
    case "PUT":
    case "DELETE":
    case "CONNECT":
    case "OPTIONS":
    case "TRACE":
    case "PATCH":
      return true;
    default:
      return false;
  }
}
