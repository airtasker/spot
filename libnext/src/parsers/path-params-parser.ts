import { ParameterDeclaration, TypeGuards } from "ts-simple-ast";
import {
  extractJsDocComment,
  extractPropertySignature
} from "./utilities/parser-utility";
import { parseType } from "./utilities/type-parser";
import { PathParamDefinition } from "../models/definitions";

/**
 * Parse an `@pathParams` decorated parameter.
 *
 * @param parameter a parameter declaration
 */
export function parsePathParams(
  parameter: ParameterDeclaration
): PathParamDefinition[] {
  // sanity check
  parameter.getDecoratorOrThrow("pathParams");
  if (parameter.hasQuestionToken()) {
    throw new Error("parameter cannot be optional");
  }
  const type = parameter.getType();
  if (!type.isObject()) {
    throw new Error("expected object");
  }
  const properties = type.getProperties();
  return properties.map(property => {
    const propertySignature = extractPropertySignature(property);
    return {
      name: propertySignature.getName(),
      description: extractJsDocComment(propertySignature),
      type: parseType(propertySignature.getType())
    };
  });
}
