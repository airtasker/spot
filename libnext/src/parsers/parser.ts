import { CompilerOptions, Project, SourceFile, ts } from "ts-simple-ast";
import { ContractNode, EndpointNode } from "../models/nodes";
import { ReferenceType } from "../models/types";
import { parseApi } from "./nodes/api-parser";
import { parseEndpoint } from "./nodes/endpoint-parser";
import { extractJsDocComment } from "./utilities/parser-utility";
import { parseInterfaceDeclaration, parseType } from "./utilities/type-parser";
import {
  retrieveTypeReferencesFromEndpoints,
  retrieveTypeReferencesFromType,
  uniqueReferences
} from "./utilities/type-reference-resolver";

export function parseFilePath(
  sourcePath: string,
  customCompilerOptions: CompilerOptions = {}
): ContractNode {
  const defaultCompilerOptions: CompilerOptions = {
    target: ts.ScriptTarget.ESNext,
    module: ts.ModuleKind.CommonJS,
    strict: true,
    noImplicitAny: true,
    strictNullChecks: true,
    strictFunctionTypes: true,
    strictPropertyInitialization: true,
    noImplicitThis: true,
    alwaysStrict: true,
    noImplicitReturns: true,
    noFallthroughCasesInSwitch: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    experimentalDecorators: true
  };

  const compilerOptions = {
    ...defaultCompilerOptions,
    ...customCompilerOptions
  };

  // Creates a new typescript program in memory
  const project = new Project({ compilerOptions });
  const sourceFile = project.addExistingSourceFile(sourcePath);

  // Add all dependent files that the project requires
  project.resolveSourceFileDependencies();

  // Validate that the project has no TypeScript syntax errors
  validateProject(project);

  return parseRootSourceFile(sourceFile, project);
}

/**
 * Parse a root source file to return a contract.
 */
function parseRootSourceFile(
  file: SourceFile,
  projectContext: Project
): ContractNode {
  const apiClasses = file
    .getClasses()
    .filter(klass => klass.getDecorator("api") !== undefined);

  if (apiClasses.length !== 1) {
    throw new Error(
      `expected class decorated with @api to be defined exactly once in root source file, found ${
        apiClasses.length
      } usages`
    );
  }
  const api = parseApi(apiClasses[0]);

  const sourceEndpoints = file
    .getClasses()
    .filter(klass => klass.getDecorator("endpoint") !== undefined)
    .map(endpointClass => parseEndpoint(endpointClass));

  const importedFiles = file
    .getImportDeclarations()
    .map(myImport => myImport.getModuleSpecifierSourceFileOrThrow());

  const endpoints = importedFiles.reduce<EndpointNode[]>(
    (endpointsAcc, currentFile) =>
      currentFile
        .getClasses()
        .filter(klass => klass.getDecorator("endpoint") !== undefined)
        .map(endpointClass => parseEndpoint(endpointClass))
        .concat(endpointsAcc),
    sourceEndpoints
  );

  // Direct reference types, filtered to unique references for efficiency
  const uniqueDirectReferenceTypes = uniqueReferences(
    retrieveTypeReferencesFromEndpoints(endpoints)
  );

  // Reference types from the direct reference type hierarchy
  const secondaryReferenceTypes = uniqueDirectReferenceTypes.reduce<
    ReferenceType[]
  >(
    (referenceTypesAcc, currentReferenceType) =>
      referenceTypesAcc.concat(
        retrieveTypeReferencesFromType(currentReferenceType, projectContext)
      ),
    []
  );

  // Combine and filter to unique type references
  const allReferenceTypes = uniqueReferences(
    uniqueDirectReferenceTypes.concat(secondaryReferenceTypes)
  );

  // Construct the equivalent TypeNode for each reference type
  const types = allReferenceTypes.map(referenceType => {
    const file = projectContext.getSourceFileOrThrow(referenceType.location);

    const typeAlias = file.getTypeAlias(referenceType.name);
    if (typeAlias !== undefined) {
      const name = typeAlias.getName();
      const description = extractJsDocComment(typeAlias);
      const type = parseType(typeAlias.getTypeNodeOrThrow());
      return { description, name, type };
    }

    const interfaceDeclaration = file.getInterface(referenceType.name);
    if (interfaceDeclaration) {
      const name = interfaceDeclaration.getName();
      const description = extractJsDocComment(interfaceDeclaration);
      const type = parseInterfaceDeclaration(interfaceDeclaration);
      return { description, name, type };
    }

    throw new Error("unexpected type reference");
  });

  return { api, endpoints, types };
}

/**
 * Validate an AST project's correctness.
 *
 * @param project an AST project
 */
function validateProject(project: Project) {
  const diagnostics = project.getPreEmitDiagnostics();
  if (diagnostics.length > 0) {
    throw new Error(
      diagnostics.map(diagnostic => diagnostic.getMessageText()).join("\n")
    );
  }
}
