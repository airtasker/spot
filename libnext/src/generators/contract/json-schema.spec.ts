import * as path from "path";
import { cleanse } from "../../cleansers/cleanser";
import {
  arrayType,
  BOOLEAN,
  booleanLiteral,
  INTEGER,
  NULL,
  NUMBER,
  numberLiteral,
  objectType,
  referenceType,
  STRING,
  stringLiteral,
  TypeKind,
  unionType
} from "../../models/types";
import { parseFilePath } from "../../parsers/parser";
import { generateJsonSchema, jsonTypeSchema } from "./json-schema";

const EXAMPLE_PATH = path.join(
  __dirname,
  "..",
  "..",
  "test",
  "examples",
  "contract.ts"
);

describe("JSON Schema generator", () => {
  test("produces valid code", async () => {
    const contractNode = await parseFilePath(EXAMPLE_PATH, {
      baseUrl: ".",
      paths: {
        "@airtasker/spotnext": ["./libnext/src/lib"]
      }
    });
    const contractDefinition = cleanse(contractNode);
    expect(generateJsonSchema(contractDefinition, "json")).toMatchSnapshot(
      "json"
    );
    expect(generateJsonSchema(contractDefinition, "yaml")).toMatchSnapshot(
      "yaml"
    );
  });

  describe("generates type validator", () => {
    test("null", () => {
      expect(jsonTypeSchema(NULL)).toMatchInlineSnapshot(`
Object {
  "type": "null",
}
`);
    });

    test("boolean", () => {
      expect(jsonTypeSchema(BOOLEAN)).toMatchInlineSnapshot(`
Object {
  "type": "boolean",
}
`);
    });

    test("boolean literal", () => {
      expect(jsonTypeSchema(booleanLiteral(true))).toMatchInlineSnapshot(`
Object {
  "const": true,
  "type": "boolean",
}
`);
      expect(jsonTypeSchema(booleanLiteral(false))).toMatchInlineSnapshot(`
Object {
  "const": false,
  "type": "boolean",
}
`);
    });

    test("string", () => {
      expect(jsonTypeSchema(STRING)).toMatchInlineSnapshot(`
Object {
  "type": "string",
}
`);
    });

    test("string literal", () => {
      expect(jsonTypeSchema(stringLiteral("some literal")))
        .toMatchInlineSnapshot(`
Object {
  "const": "some literal",
  "type": "string",
}
`);
    });

    test("number", () => {
      expect(jsonTypeSchema(NUMBER)).toMatchInlineSnapshot(`
Object {
  "type": "number",
}
`);
    });

    test("number literal", () => {
      expect(jsonTypeSchema(numberLiteral(1.5))).toMatchInlineSnapshot(`
Object {
  "const": 1.5,
  "type": "number",
}
`);
      expect(jsonTypeSchema(numberLiteral(-23.1))).toMatchInlineSnapshot(`
Object {
  "const": -23.1,
  "type": "number",
}
`);
    });

    test("integer", () => {
      expect(jsonTypeSchema(INTEGER)).toMatchInlineSnapshot(`
Object {
  "type": "integer",
}
`);
    });

    test("number literal", () => {
      expect(jsonTypeSchema(numberLiteral(0))).toMatchInlineSnapshot(`
Object {
  "const": 0,
  "type": "integer",
}
`);
      expect(jsonTypeSchema(numberLiteral(123))).toMatchInlineSnapshot(`
Object {
  "const": 123,
  "type": "integer",
}
`);
      expect(jsonTypeSchema(numberLiteral(-1000))).toMatchInlineSnapshot(`
Object {
  "const": -1000,
  "type": "integer",
}
`);
    });

    test("object", () => {
      expect(jsonTypeSchema(objectType([]))).toMatchInlineSnapshot(`
Object {
  "properties": Object {},
  "required": Array [],
  "type": "object",
}
`);
      expect(
        jsonTypeSchema(
          objectType([
            {
              name: "singleField",
              type: NUMBER,
              optional: false
            }
          ])
        )
      ).toMatchInlineSnapshot(`
Object {
  "properties": Object {
    "singleField": Object {
      "type": "number",
    },
  },
  "required": Array [
    "singleField",
  ],
  "type": "object",
}
`);
      expect(
        jsonTypeSchema(
          objectType([
            {
              name: "field1",
              type: NUMBER,
              optional: false
            },
            {
              name: "field2",
              type: STRING,
              optional: false
            },
            {
              name: "field3",
              type: BOOLEAN,
              optional: true
            }
          ])
        )
      ).toMatchInlineSnapshot(`
Object {
  "properties": Object {
    "field1": Object {
      "type": "number",
    },
    "field2": Object {
      "type": "string",
    },
    "field3": Object {
      "type": "boolean",
    },
  },
  "required": Array [
    "field1",
    "field2",
  ],
  "type": "object",
}
`);
    });

    test("array", () => {
      expect(jsonTypeSchema(arrayType(STRING))).toMatchInlineSnapshot(`
Object {
  "items": Object {
    "type": "string",
  },
  "type": "array",
}
`);
    });

    test("union", () => {
      expect(jsonTypeSchema(unionType([STRING, NUMBER, BOOLEAN])))
        .toMatchInlineSnapshot(`
Object {
  "oneOf": Array [
    Object {
      "type": "string",
    },
    Object {
      "type": "number",
    },
    Object {
      "type": "boolean",
    },
  ],
}
`);
    });

    test("type reference", () => {
      expect(
        jsonTypeSchema(referenceType("OtherType", "location", TypeKind.STRING))
      ).toMatchInlineSnapshot(`
Object {
  "$ref": "#/definitions/OtherType",
}
`);
    });
  });
});
