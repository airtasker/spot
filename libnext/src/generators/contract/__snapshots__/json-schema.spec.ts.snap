// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON Schema generator produces valid code: json 1`] = `
"{
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"definitions\\": {
    \\"UserBody\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"data\\": {
          \\"type\\": \\"object\\",
          \\"properties\\": {
            \\"firstName\\": {
              \\"type\\": \\"string\\"
            },
            \\"lastName\\": {
              \\"type\\": \\"string\\"
            },
            \\"profile\\": {
              \\"$ref\\": \\"#/definitions/Profile\\"
            }
          },
          \\"required\\": [
            \\"firstName\\",
            \\"lastName\\",
            \\"profile\\"
          ]
        }
      },
      \\"required\\": [
        \\"data\\"
      ]
    },
    \\"ErrorBody\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\"
        },
        \\"message\\": {
          \\"type\\": \\"array\\",
          \\"items\\": {
            \\"type\\": \\"string\\"
          }
        }
      },
      \\"required\\": [
        \\"name\\",
        \\"message\\"
      ]
    },
    \\"CreateUserRequestBody\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"data\\": {
          \\"type\\": \\"object\\",
          \\"properties\\": {
            \\"firstName\\": {
              \\"type\\": \\"string\\"
            },
            \\"lastName\\": {
              \\"type\\": \\"string\\"
            },
            \\"age\\": {
              \\"type\\": \\"number\\"
            },
            \\"email\\": {
              \\"$ref\\": \\"#/definitions/Email\\"
            },
            \\"address\\": {
              \\"$ref\\": \\"#/definitions/Address\\"
            }
          },
          \\"required\\": [
            \\"firstName\\",
            \\"lastName\\",
            \\"age\\",
            \\"email\\",
            \\"address\\"
          ]
        }
      },
      \\"required\\": [
        \\"data\\"
      ]
    },
    \\"Profile\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"private\\": {
          \\"type\\": \\"boolean\\"
        },
        \\"messageOptions\\": {
          \\"$ref\\": \\"#/definitions/MessageOptions\\"
        }
      },
      \\"required\\": [
        \\"private\\",
        \\"messageOptions\\"
      ]
    },
    \\"MessageOptions\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"newsletter\\": {
          \\"type\\": \\"boolean\\"
        }
      },
      \\"required\\": [
        \\"newsletter\\"
      ]
    },
    \\"Email\\": {
      \\"type\\": \\"string\\"
    },
    \\"Address\\": {
      \\"type\\": \\"string\\"
    }
  }
}"
`;

exports[`JSON Schema generator produces valid code: yaml 1`] = `
"$schema: 'http://json-schema.org/draft-07/schema#'
definitions:
  UserBody:
    type: object
    properties:
      data:
        type: object
        properties:
          firstName:
            type: string
          lastName:
            type: string
          profile:
            $ref: '#/definitions/Profile'
        required:
          - firstName
          - lastName
          - profile
    required:
      - data
  ErrorBody:
    type: object
    properties:
      name:
        type: string
      message:
        type: array
        items:
          type: string
    required:
      - name
      - message
  CreateUserRequestBody:
    type: object
    properties:
      data:
        type: object
        properties:
          firstName:
            type: string
          lastName:
            type: string
          age:
            type: number
          email:
            $ref: '#/definitions/Email'
          address:
            $ref: '#/definitions/Address'
        required:
          - firstName
          - lastName
          - age
          - email
          - address
    required:
      - data
  Profile:
    type: object
    properties:
      private:
        type: boolean
      messageOptions:
        $ref: '#/definitions/MessageOptions'
    required:
      - private
      - messageOptions
  MessageOptions:
    type: object
    properties:
      newsletter:
        type: boolean
    required:
      - newsletter
  Email:
    type: string
  Address:
    type: string
"
`;
